### PR TITLE
feat(file-based): throttle legacy per-slice state emission by default

### PR DESCRIPTION
**Pull request by tolik0**
## What

File-based streams now emit at most one per-slice STATE message per 10 minutes on the legacy (non-concurrent) read path. The first slice and the final state of the sync always emit, so destinations still see the latest cursor — only intermediate checkpoint granularity is reduced.

## Why

On the legacy `DefaultFileBasedCursor` path the framework emits a state message after every slice, and a slice is a single file. The payload carries the entire file-history dict, so the cost grows with the sync: measured at **1.72 MB per message and ~2.57 GB per attempt** on source-gcs (oncall#13210, and oncall#12663 before it on source-s3). The platform buffers un-ACKed state messages, so the replication pod dies before the destination catches up.

## Where it lives

The throttle is a `CheckpointReader` decorator. **`Stream.read()` and `test_stream_read.py` are byte-identical to `main`** — nothing is added to the generic `Stream` base class.

Suppression rides on the documented contract rather than a new mechanism. From `CheckpointReader.get_checkpoint()`:

> *"Retrieves the current state value of the stream. The connector does not emit state messages if the checkpoint value is None."*

`CursorBasedCheckpointReader` already uses exactly that to suppress duplicate states.

```python
# ThrottledCheckpointReader.get_checkpoint()
if self._finished:                       # final call, after iteration ended
    return checkpoint or self._pending   # pay out anything suppressed
if now - self._last_surfaced_at < self._throttle_seconds:
    self._pending = checkpoint           # hold it
    return None                          # -> caller emits nothing
self._last_surfaced_at = now
return checkpoint
```

`DefaultFileBasedStream._get_checkpoint_reader()` wraps whatever reader the base class picks, so that choice stays with the base class even if it changes later:

```python
def _get_checkpoint_reader(self, *args, **kwargs):
    reader = super()._get_checkpoint_reader(*args, **kwargs)
    if self.state_emission_throttle_seconds is None:
        return reader
    return ThrottledCheckpointReader(reader, self.state_emission_throttle_seconds)
```

Setting `state_emission_throttle_seconds = None` returns the unwrapped reader, so the throttle can be turned off without a separate code path.

A literal cursor-only implementation is not possible: `AbstractFileBasedStream.get_cursor()` returns `None` by design (*"the file-based cursor isn't compatible with the cursor-based iteration flow in core.py"*), so `DefaultFileBasedCursor` sits outside the checkpoint machinery and has no hook that runs at emission time.

## Design notes

**Not configurable, and not opt-in.** An earlier revision of this PR exposed `FileBasedSource._stream_state_emission_throttle_seconds` so each connector could pick a value. That was wrong twice over:

1. **The value has no per-connector meaning.** 600 s is not a platform limit — I checked. The only platform timer watching source output is `heartbeat-max-seconds-between-messages` (**10800 s / 3 h**), and `LocalContainerAirbyteSource.shouldBeat()` resets it on `RECORD` as well as `STATE`, so throttling state cannot stall a sync that is still producing records. The 600 s figure traces back to `ConcurrentPerPartitionCursor`, which hardcodes it. This PR extracts that literal into a shared `DEFAULT_STATE_EMISSION_THROTTLE_SECONDS` and, because sharing only the constant still left the two comparisons free to drift, a shared `state_emission_is_due()` predicate that both paths call. The boundary is inclusive — what "at most once every N seconds" means — which corrects the concurrent path from `elapsed > N` to `elapsed >= N`. A regression test drives both call sites at the same elapsed time.

2. **A source-level hook could not reach the connectors it was written for.** `source-s3` and `source-gcs` both override `_make_default_stream()` without calling `super()`, so anything the source injected was invisible to them. Owning the reader on the stream class reaches their subclasses (`GCSStream`, `ThrottledFileBasedStream`) and `PermissionsFileBasedStream` by plain inheritance — there is nothing left to bypass.

**Blast radius is bounded to legacy file-based streams.** The concurrent path is untouched: `FileBasedStreamFacade` overrides `read()` and never reaches `Stream.read()`.

## Tradeoff

A failed sync re-reads up to 10 minutes of files, because the last acknowledged checkpoint can be that stale. This is a real regression in recovery granularity, accepted deliberately: an unbounded state buffer kills the sync outright, while a coarser checkpoint only costs rework.

## Test changes

`test_file_based_read[multi_csv_remove_old_files_if_history_is_full]` asserted one state per file. Its middle state is now suppressed by the throttle, so the expectation drops that entry — the cursor still advances through `b.csv`, carried in the final state. This is the **only** existing test whose expectations changed.

New coverage:

- `unit_tests/sources/streams/checkpoint/test_throttled_checkpoint_reader.py` — suppress-and-force-final, periodic re-emit once the window elapses, no duplicate emit when the last slice lands exactly on the boundary, a non-positive window failing open, and the inner reader's own final checkpoint taking precedence. Each asserts the full emitted cursor sequence, not just the message count. Driven against the reader directly with an injected clock, so there is no global `time.monotonic` patching and no stream setup.
- `unit_tests/sources/file_based/stream/test_state_emission_throttle.py` — the stream wires the reader up, a connector-style subclass cannot bypass it, and `None` restores the unwrapped reader.

## Follow-ups

- **source-gcs**: bump the CDK. No connector-side code change needed — inheritance delivers the throttle.
- **source-s3**: delete `source_s3/v4/throttled_stream.py` and its `_make_default_stream` override. Until then s3 double-throttles at the same 600 s window, which is harmless (spacing is still ≥600 s and the final state still emits) but redundant.
- This bounds state *volume*. It does **not** fix the re-read loop in oncall#13210, whose cause is history-size eviction in `DefaultFileBasedCursor`; that is tracked separately and mitigated by `days_to_sync_if_history_is_full`.

## CI note

`destination-motherduck` and `source-google-drive` are red for reasons unrelated to this PR. motherduck fails on every open PR. google-drive fails in test *setup* — the standard-test harness calls `connector()` with no arguments (`connector_base.py:94`) while `SourceGoogleDrive.__init__` requires `catalog`, `config`, and `state`, so every test including `test_spec` dies with a `TypeError` before any test body runs. It has failed all 6 times it has ever executed, including on unrelated branches predating this PR; it only surfaces here because `cdk_extra: file-based` skips it elsewhere. Same defect already commented out for `source-zendesk-support` in the workflow matrix.

## Summary by CodeRabbit

* **New Features**
  * Added throttling for checkpoint and state emissions, reducing intermediate updates while preserving the latest final state.
  * Standardized the default emission interval across incremental and file-based streams.
  * Added an option to disable throttling where needed.

* **Bug Fixes**
  * State emissions now occur correctly at the throttle boundary and on the first or final checkpoint.

* **Tests**
  * Added coverage for timing boundaries, stream inheritance, disabled throttling, delegation, and final-state behavior.

### CODE REVIEW

File: airbyte_cdk/sources/streams/core.py [hunk 1]
```diff
@@ -240,17 +268,38 @@ def read(  # type: ignore  # ignoring typing for ConnectorStateManager because o
             self._observe_state(checkpoint_reader)
             checkpoint_state = checkpoint_reader.get_checkpoint()
             if should_checkpoint and checkpoint_state is not None:
-                airbyte_state_message = self._checkpoint_state(
-                    checkpoint_state, state_manager=state_manager
-                )
-                yield airbyte_state_message
+                last_observed_checkpoint = checkpoint_state
```
[the remark this answers is not in the document]
Comment (tolik0) [reply airbyte_cdk/sources/streams/core.py]: Fixed in 5a99392 — the assignment was a leftover from collapsing `last_observed_checkpoint` and `throttled_pending_emit` into a single `pending_checkpoint` local. It was genuinely dead: nothing reads it, and the suppressed checkpoint is carried by `pending_checkpoint` instead. Removed.

Worth noting Ruff didn't catch this — F841 isn't in the repo's enabled rule set, so the lint gate passed on it.

File: airbyte_cdk/sources/streams/core.py [hunk 2]
```diff
@@ -184,6 +194,16 @@ def read(  # type: ignore  # ignoring typing for ConnectorStateManager because o
...
         stream_state_tracker = copy.deepcopy(stream_state)
+
+        # State-emission throttle bookkeeping. Active only when the stream sets
+        # `state_emission_throttle_seconds`; default behavior is preserved.
+        # `None` means nothing has been emitted yet, so the first per-slice
+        # checkpoint always fires regardless of the clock's absolute value.
+        throttle_seconds = self.state_emission_throttle_seconds
+        last_state_emit_at: Optional[float] = None
+        last_observed_checkpoint: Optional[Mapping[str, Any]] = None
+        throttled_pending_emit = False
```
[the remark this answers is not in the document]
Comment (tolik0) [reply airbyte_cdk/sources/streams/core.py]: Good catch on the ambiguity, though I went with documenting over validating.

A non-positive value doesn't silently defeat the feature — it degrades to exactly the current default. The window check is `elapsed < throttle_seconds`, which is never true for `0` or a negative value, so every slice emits just as it does on `main`. The failure mode is "no throttling", which is the historical behaviour and the safe direction to fail in.

Raising felt disproportionate for a value that's harmless, and clamping to `None` is already what effectively happens. This is also a code-level class attribute set by connector authors, not user-supplied config, so there's no runtime misconfiguration surface to guard.

Added a docstring note in 47bf615 so the intent is explicit rather than incidental.

File: airbyte_cdk/sources/file_based/file_based_source.py [hunk 3]
```diff
@@ -349,6 +361,9 @@ def _make_default_stream(
             use_file_transfer=use_file_transfer(parsed_config),
             preserve_directory_structure=preserve_directory_structure(parsed_config),
         )
+        if self._stream_state_emission_throttle_seconds is not None:
+            stream.state_emission_throttle_seconds = self._stream_state_emission_throttle_seconds
+        return stream
```
[the remark this answers is not in the document]
Comment (tolik0) [reply airbyte_cdk/sources/file_based/file_based_source.py]: Agreed, and fixed in a9bd031 — this was the same root problem a human reviewer flagged as blocking, from the other direction.

The placement was worse than just missing the permissions path: `source-s3` and `source-gcs` both override `_make_default_stream` **without** calling `super()`, so the hook was unreachable for the two connectors the feature was written for. The planned `source-gcs` follow-up would have landed as a silent no-op — no error, no log, just an unthrottled sync.

Moved the assignment into `_make_file_based_stream()` behind a small `_apply_state_emission_throttle()` helper, exactly as you suggest. That's the single funnel both factories pass through, nothing overrides it, and it picks up `_make_permissions_stream` for free.

Also added `unit_tests/sources/file_based/test_state_emission_throttle_propagation.py`, including a subclass that overrides `_make_default_stream` without `super()` to mirror the two connectors — that test fails against the old placement.

**Comment by tolik0** [comment]
Fixed in fc0caf3e. Copilot raised this on two consecutive commits and it was right both times — the part that was actually wrong was my PR description claiming the two paths were in sync.

They shared the constant but not the comparison:

- `ConcurrentPerPartitionCursor`: `elapsed <= throttle` suppresses → only `elapsed > N` emits
- `ThrottledCheckpointReader`: `elapsed < throttle` suppresses → `elapsed == N` emits

So at exactly the window they disagreed. Sharing a constant was never going to prevent that, since the comparison lived at each call site.

Extracted `state_emission_is_due(last_emitted_at, now, throttle_seconds)` and routed both through it, so consistency is structural instead of conventional. The boundary is inclusive, which is what "at most once every N seconds" means, so the concurrent path moves from `>` to `>=`. With wall-clock floats an exact hit is vanishingly unlikely, so this is a semantic correction rather than an observable behaviour change — and no existing test pinned the old boundary.

The helper also absorbs two edge cases that were previously encoded separately at each site: `None` meaning "nothing emitted yet, always due" (the reader), and the `0.0` sentinel compared against an epoch clock (the cursor, which initialises `_last_emission_time = 0.0` and uses `time.time()`). A non-positive window still fails open to unthrottled.

Coverage: parametrized cases for the predicate — inside, exactly on, and past the boundary, the epoch-vs-sentinel cold start, and zero/negative windows — plus `test_both_throttle_paths_agree_on_the_boundary`, which drives the real `_throttle_state_message` and the reader at the same elapsed time. I confirmed that test fails when the cursor's `<=` is restored, so it is a genuine guard rather than one that passes either way.

PR description updated to match.

File: airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py [hunk 4]
```diff
@@ -17,6 +18,39 @@ class CheckpointMode(Enum):
...
+    exactly N seconds after the previous one is due.
+
+    `last_emitted_at is None` means nothing has been emitted yet, so the first
+    emission is always due regardless of the clock's absolute value. A
+    `throttle_seconds <= 0` never suppresses, which degrades to the historical
+    unthrottled behaviour rather than erroring.
+    """
+    if last_emitted_at is None:
+        return True
+    return now - last_emitted_at >= throttle_seconds
```
[the remark this answers is not in the document]
Comment (tolik0) [reply airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py]: Good catch — fixed in 769753f2. This was a real contract violation, not just a hardening suggestion: the docstring promises a non-positive window never suppresses, and the implementation didn't deliver that.

Reproduced before fixing:

```
state_emission_is_due(last=1000.0, now=990.0, throttle=0.0)   -> False
state_emission_is_due(last=1000.0, now=990.0, throttle=-5.0)  -> False
```

And you're right that it's reachable rather than theoretical — `ConcurrentPerPartitionCursor` measures with wall-clock `time.time()`, which NTP can step backwards. The reader defaults to `time.monotonic()`, but the clock is injectable, so the predicate shouldn't rely on the caller's choice.

Took your diff as proposed. I also pinned the neighbouring case explicitly — a *positive* window still suppresses under a rollback, which is correct since the window genuinely hasn't elapsed — so the guard can't be widened past `<= 0` by accident later.

File: unit_tests/sources/file_based/stream/test_state_emission_throttle.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    assert isinstance(_reader(_stream(_ConnectorStream)), ThrottledCheckpointReader)
+
+
+def test_permissions_stream_is_throttled_too() -> None:
+    """The permissions transfer path is on the same legacy emission path."""
+    assert (
+        PermissionsFileBasedStream.state_emission_throttle_seconds
+        == DEFAULT_STATE_EMISSION_THROTTLE_SECONDS
+    )
```
[the remark this answers is not in the document]
Comment (tolik0) [reply unit_tests/sources/file_based/stream/test_state_emission_throttle.py:96]: Agreed — fixed in 769753f2. The test was strictly weaker than its two siblings, which both assert the constructed reader; this one only checked the inherited attribute, so it would have gone green while the throttling path was gone.

It now constructs the stream and asserts the reader:

```python
stream = _stream(PermissionsFileBasedStream, stream_permissions_reader=MagicMock())
assert isinstance(_reader(stream), ThrottledCheckpointReader)
```

Kept the class-level assertion alongside it, since the two failures mean different things: a wrong window versus no throttle at all.

File: unit_tests/sources/streams/checkpoint/test_throttled_checkpoint_reader.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+            1000.0,
+            1600.0,
+            600.0,
+            True,
+            "'once every N seconds' makes the boundary inclusive",
+            id="exactly_on_boundary",
+        ),
+        pytest.param(1000.0, 1600.1, 600.0, True, "past the window", id="past"),
+        pytest.param(1000.0, 1000.0, 0.0, True, "zero never suppresses", id="zero_fails_open"),
+        pytest.param(1000.0, 1000.0, -5.0, True, "negative never suppresses", id="negative"),
```
[the remark this answers is not in the document]
Comment (tolik0) [reply unit_tests/sources/streams/checkpoint/test_throttled_checkpoint_reader.py:186]: Both done in 769753f2 — the predicate now returns `True` before any arithmetic when `throttle_seconds <= 0`, and `state_emission_is_due(1000.0, 990.0, -5.0)` is covered along with the zero-throttle equivalent.

I checked these are genuine guards rather than tests that pass either way: reverting the predicate to its previous form fails exactly the two new cases —

```
FAILED test_state_emission_is_due[zero_clock_rollback]     - zero, clock stepped backwards
FAILED test_state_emission_is_due[negative_clock_rollback] - negative, clock stepped backwards
```

and restoring the fix returns 19 passed.

Added a third case for contrast: a *positive* window under the same rollback still suppresses, which is the correct behaviour and stops the `<= 0` guard being loosened later.

**Comment by tolik0** [comment]
/prerelease
> **Prerelease Job Info**
>
> This job triggers the publish workflow with default arguments to create a prerelease.
>
> Prerelease job started... [Check job output.][1]

[1]: https://github.com/airbytehq/airbyte-python-cdk/actions/runs/31038474743
> ✅ Prerelease workflow triggered successfully.
>
> View the publish workflow run: https://github.com/airbytehq/airbyte-python-cdk/actions/runs/31038498812

**Review by darynaishchenko — APPROVED** [review]
